# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
# Libraries
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
from pandas.plotting import lag_plot
from statsmodels.tsa.stattools import adfuller
import warnings

# Setup
warnings.filterwarnings("ignore")

# Challenge 1 - Loading and Evaluating The Data

In this lab, we will look at a dataset of sensor data from a cellular phone. The phone was carried in the subject's pocket for a few minutes while they walked around.

To load the data, run the code below.

In [2]:
# Read data
data=pd.read_csv(r'/Users/akbare/Desktop/IronhackDA/DAFT_212/module_2/Lab_20_Time-Series-Analysis/sub_1.csv')



Examine the data using the `head` function.

In [3]:
data.head()

Unnamed: 0  attitude.roll  attitude.pitch  attitude.yaw  gravity.x  \
0           0       1.528132       -0.733896      0.696372   0.741895   
1           1       1.527992       -0.716987      0.677762   0.753099   
2           2       1.527765       -0.706999      0.670951   0.759611   
3           3       1.516768       -0.704678      0.675735   0.760709   
4           4       1.493941       -0.703918      0.672994   0.760062   

   gravity.y  gravity.z  rotationRate.x  rotationRate.y  rotationRate.z  \
0   0.669768  -0.031672        0.316738        0.778180        1.082764   
1   0.657116  -0.032255        0.842032        0.424446        0.643574   
2   0.649555  -0.032707       -0.138143       -0.040741        0.343563   
3   0.647788  -0.041140       -0.025005       -1.048717        0.035860   
4   0.647210  -0.058530        0.114253       -0.912890        0.047341   

   userAcceleration.x  userAcceleration.y  userAcceleration.z  
0            0.294894           -0.184493            0.377542  
1            0.219405            0.035846            0.114866  
2            0.010714            0.134701           -0.167808  
3           -0.008389            0.136788            0.094958  
4            0.199441            0.353996           -0.044299

Check whether there is any missing data. If there is any missing data, remove the rows containing missing data.

In [4]:
data.isnull()

Unnamed: 0  attitude.roll  attitude.pitch  attitude.yaw  gravity.x  \
0          False          False           False         False      False   
1          False          False           False         False      False   
2          False          False           False         False      False   
3          False          False           False         False      False   
4          False          False           False         False      False   
...          ...            ...             ...           ...        ...   
1746       False          False           False         False      False   
1747       False          False           False         False      False   
1748       False          False           False         False      False   
1749       False          False           False         False      False   
1750       False          False           False         False      False   

      gravity.y  gravity.z  rotationRate.x  rotationRate.y  rotationRate.z  \
0         False      False           False           False           False   
1         False      False           False           False           False   
2         False      False           False           False           False   
3         False      False           False           False           False   
4         False      False           False           False           False   
...         ...        ...             ...             ...             ...   
1746      False      False           False           False           False   
1747      False      False           False           False           False   
1748      False      False           False           False           False   
1749      False      False           False           False           False   
1750      False      False           False           False           False   

      userAcceleration.x  userAcceleration.y  userAcceleration.z  
0                  False               False               False  
1                  False               False               False  
2                  False               False               False  
3                  False               False               False  
4                  False               False               False  
...                  ...                 ...                 ...  
1746               False               False               False  
1747               False               False               False  
1748               False               False               False  
1749               False               False               False  
1750               False               False               False  

[1751 rows x 13 columns]

How many rows and columns are in our data?

In [12]:
data.shape

(1751, 13)

To perform time series analysis on the data, we must change the index from a range index to a time series index. In the cell below, create a time series index using the `pd.date_range` function. Create a time series index starting at 1/1/2018 00:00:00 and ending at 1/1/2018 00:29:10. The number of periods is equal to the number of rows in `sensor`. The frequency should be set to `infer`.

In [ ]:
index=pd.date_range(start='1/1/2018 00:00:00', end='1/1/2018 00:29:10', periods=1751)

Assign the time series index to the dataframe's index.

In [ ]:
import statsmodels.api as sm
from statsmodels.tsa.stattools import adfuller

Our next step is to decompose the time series and evaluate the patterns in the data. Load the `statsmodels.api` submodule and plot the decomposed plot of `userAcceleration.x`. Set `freq=60` in the `seasonal_decompose` function. Your graph should look like the one below.

![time series decomposition](./images/tsa_decompose.png)

In [ ]:
result= sm.tsa.seasonal_decompose(sensor['userAcceleration.x'], freq=60)
result.plot()

Plot the decomposed time series of `rotationRate.x` also with a frequency of 60.

In [ ]:
result2= sm.tsa.seasonal_decompose(sensor['rotationRate.x'], freq=60)
result2.plot()

# Challenge 2 - Modelling the Data

To model our data, we should look at a few assumptions. First, let's plot the `lag_plot` to detect any autocorrelation. Do this for `userAcceleration.x`.

In [ ]:
lag_plot(sensor['userAcceleration.x'])
plt.show()

Create a lag plot for `rotationRate.x`

In [ ]:
lag_plot(sensor['rotationRate.x'])
plt.show()

What are your conclusions from both visualizations?

```Comments:``` Autocorrelation in time series means that the series is linearly related to a lagged version of itself. For features 'rotationRate.x' and 'userAcceleration.x', the observations at time *t* have a positive linear correlation with the observations at time *t+1*. One of the assumptions of regression analysis is that the data has no autocorrelation. If we try to do a regression analysis on data with autocorrelation, then our analysis will be misleading.

The next step will be to test both variables for stationarity. Perform the Augmented Dickey Fuller test on both variables below.

In [ ]:
# Ho: userAcceleration.x is not stationary
# Ha: userAcceleration.x is stationary


In [ ]:
from statsmodels.tsa.stattools import adfuller

result = adfuller(sensor['rotationRate.x'])
print('ADF Test Statistic: %.2f' % result[0])
print('5%% Critical Value: %.2f' % result[4]['5%'])
print('p-value: %.2f' % result[1])

What are your conclusions from this test?

```Comments:``` ARMA models work on the assumption that the time series is stationary. Stationarity means that the statistical properties of a process generating a time series do not change over time. It does not mean that the series does not change over time, just that the way it changes does not itself change over time. In both cases, we can reject the null hypothesis that the time series is not stationary, so we can create an ARMA model.

Finally, we'll create an ARMA model for `userAcceleration.x`. Load the `ARMA` function from `statsmodels`. The order of the model is (2, 1). Split the data to train and test. Use the last 10 observations as the test set and all other observations as the training set. 

In [ ]:
from statsmodels.tsa.arima_model import ARMA
# ARMA of order (p,q)
# p: order of the autoregressive model (AR).
# q: order of the moving average model (MA).

In [ ]:
index = sensor.shape[0] - 10
train = sensor['userAcceleration.x'][:index] 
test = sensor['userAcceleration.x'][index:]
print(f"The number of samples in train is {train.shape[0]} and the number of samples in test is {test.shape[0]}."

In [ ]:
# Model
model = ARMA(train, order=(2,1), freq='S')
arma = model.fit(disp=False)

# Predictions
predictions = arma.predict(test.index[0], test.index[-1])

To compare our predictions with the observed data, we can compute the RMSE (Root Mean Squared Error) from the submodule `statsmodels.tools.eval_measures`. You can read more about this function [here](https://www.statsmodels.org/dev/generated/statsmodels.tools.eval_measures.rmse.html). Compute the RMSE for the last 10 rows of the data by comparing the observed and predicted data for the `userAcceleration.x` column.

In [ ]:
from statsmodels.tools.eval_measures import rmse

In [ ]:
rmse(test, predictions)

```Comments:``` 
The RMSE is the square root of the variance of the residuals. It indicates the absolute fit of the model to the data, so how close the observed data points are to the model’s predicted values. Whereas R-squared is a relative measure of fit, RMSE is an absolute measure of fit. As the square root of a variance, RMSE can be interpreted as the standard deviation of the unexplained variance, and has the useful property of being in the same units as the response variable. Lower values of RMSE indicate better fit. RMSE is a good measure of how accurately the model predicts the response, and it is the most important criterion for fit if the main purpose of the model is prediction. We can say that the deviation from the true prediction is +/- 0.22.